proposals/0022-multi-stake.md

```diff
@@ -0,0 +1,181 @@
+impacts the specific implementation in the Solana Labs validator.
+
+In the perfect situation, small stakes are properly delegated, but don't receive
+rewards. Meaning:
+
+* they don't take up space in the stakes cache
+* they take the correct amount of epoch boundaries to activate / deactivate
+* they are included in a validator's voting power
+* they can be slashed
+
+There are a few ways of solving this at the stakes cache level, and they all have
+serious issues.
+
+* Include small stakes in the cache, but don't pay out rewards
+
+This is the simplest solution, which essentially changes "minimum delegation amount"
+to mean "minimum reward earning amount". So you can delegate stake accounts with
+less than the minimum delegation, but they won't earn rewards.
+
+Unfortunately, small stakes can still bloat the stakes cache from a memory level.
+
+* Don't include small stakes in the cache, but track them in transactions
+
+In this solution, on every transaction, if the runtime sees a small stake, it
+includes its delegation amount in the validator's voting power, but doesn't
+add its pubkey and `Delegation` to the cache.
+
+The runtime tracks the pre and post state of every account, and for a successful
+transaction, subtracts the old small stake delegation amount, and adds the new small
+stake delegation amount.
+
+During rewards payout, small stakes are completely omitted, since they are not
+present in the cache.
+
+During a slashing event, it's still possible to query it using `get_program_accounts`.
+
+This is incredibly brittle, however, and introduces more overhead in the runtime.
+The bank must hold onto pre-states for all transactions to debit the stakes cache. And,
+if called from the outside, all `store_accounts` functions on the bank could easily
+invalidate the stakes cache.
+
+For example, if you store a small stake directly, what does the cache do? What
+was the pre-state of that small stake? There's no way to know.
+
+## Detailed Design
+
+Rather than hacking the validator, let's change the stake program.
+
+With a new "multi stake" account, carrying *two* more `Delegation`s, the stake
+program becomes much more flexible.
+
+During initial delegation, the "multi stake" account must have enough to cover
+the minimum stake delegation amount. The two other `Delegation` instances may be
+used for small stake movements.
+
+Here are some example uses:
+
+### Add or remove small stake
+
+To add a small stake to your account, transfer the lamports, call `delegate` once
+more, and now your stake account has two delegations: the initial one, and the
+newly activating one.
+
+Once the new `Delegation` is active, a permissionless `update` instruction follows
+the existing `merge` logic to consolidate the two `Delegation`s into one.
+
+The same roughly applies for removing a small stake. You `deactivate` a portion
+of it, which adds another delegation to your stake account: the initial one, and
+the newly deactivating one.
+
+### Redelegate
```

Comment:
It is still generally expected that all `Delegations` in a multistake account be to the same validator? The only time this is false would be during a redelegation maneuver?

Reply:
Yeah, that's my general thinking. My overall idea is to restrict it a lot to start, and then we can consider opening up more possibilities. For example, if someone wants to do a partial deactivation and a redelegation in one account, we can make a change to allow that case.

Reply:
I mean we could allow N `Delegations`, where there's some reasonable N. N = 3 for now perhaps, but we keep the instructions and implementation flexible such that if we decide a larger N is suitable then it's not a big deal to extend.

But yes, restrictive to start with for sure. 

Another interesting real-world usage of multistake is the easier activation of additional lamports that are transferred into an existing stake account. Jito does this for MEV rewards rn, and it sucks because now the user needs to withdraw those rewards, make a new stake account, ...